### PR TITLE
Deprecate `OS::Mac` on Linux

### PR DESCRIPTION
**Pull request by Rylan12**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Closes https://github.com/Homebrew/brew/issues/15343

This PR deprecates the `OS::Mac` methods that are currently defined on Linux. I've gone through each instance that I can find where one of the methods was exposed to Linux code and refactored it. I'm marking this as a draft for now to gauge whether this is the right approach, and I will do more thorough testing.

### CODE REVIEW

**Review by MikeMcQuaid — COMMENTED** [review]
Nice work! This approach makes a lot of sense to me. As failing CI indicates: there's at least one formula and `brew test-bot` that will need updated before this can be merged.

Also, good to keep everything here for now for testing but may want to merge this with the `odeprecated` commented out once this is all 🟢 so it can get some testing before we do the actual deprecation before a new major or minor release.

File: Library/Homebrew/extend/os/mac/utils/curl.rb
Comment (Bo98) [inline Library/Homebrew/extend/os/mac/utils/curl.rb:12]: I reckon this can go since we don't allow system curl on El Capitan anymore.

Comment (MikeMcQuaid) [reply Library/Homebrew/extend/os/mac/utils/curl.rb:12]: Good catch @Bo98!

**Comment by Rylan12** [comment]
I've opened a test-bot PR to fix that issue. The `lsyncd` one is a little more tricky.

One possible way to fix it would be what I've done in [this commit](https://github.com/Homebrew/homebrew-core/commit/897ae54571c2004ae5ec15c899c6cf3bab1d1af9).

Since the offending lines only deal with `resource` blocks which aren't currently included in the API, not having accurate information doesn't hurt us. I don't like it though since if `resource` blocks get added in the future, it wouldn't work. I'd love to get some other thoughts on the best way to handle this.

Another possibility is to simply add a `rescue` to catch the method-undefined error like this:

```ruby
begin
  macos_version = MacOS.full_version.major_minor # Ignore bugfix/security updates

  on_catalina :or_older do
    macos_version = MacOS.full_version
  end
rescue NameError
  macos_version = nil
end
```

This avoids the `# rubocop:{dis,en}able` so maybe is better... thoughts?

**Comment by Bo98** [comment]
We likely don't need `MacOS.full_version` in `lsyncd` anymore, so it could be a series of `on_el_capitan`, `on_sierra`, etc.

**Comment by Rylan12** [comment]
> We likely don't need `MacOS.full_version` in `lsyncd` anymore, so it could be a series of `on_el_capitan`, `on_sierra`, etc.

The reason I think we still need it is that the minor version affects which resource is downloaded (e.g. `13.1` it uses version `8792.61.2` and on `13.0` it uses version `8792.41.9`). I don't know anything about `lsyncd` itself to know if it could be done differently to avoid this

**Comment by Bo98** [comment]
Technically speaking yes. Though we only provide one bottle per major OS in the end and there's never been any issues there.

**Comment by MikeMcQuaid** [comment]
All CI is 🟢.

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:58]:
```suggestion
      # odeprecated "`MacOS.version` on Linux"
```

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:63]:
```suggestion
      # odeprecated "`MacOS.full_version` on Linux"
```

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:68]:
```suggestion
      # odeprecated "`MacOS.languages` on Linux"
```

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:73]:
```suggestion
      # odeprecated "`MacOS.language` on Linux"
```

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:78]:
```suggestion
      # odeprecated "`MacOS.sdk_root_needed?` on Linux"
```

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:83]:
```suggestion
      # odeprecated "`MacOS.sdk_path_if_needed` on Linux"
```

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:88]:
```suggestion
      # odeprecated "`MacOS.sdk_path` on Linux"
```

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:94]:
```suggestion
        # odeprecated "`MacOS::Xcode.version` on Linux"
```

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:99]:
```suggestion
        # odeprecated "`MacOS::Xcode.installed?` on Linux"
```

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:106]:
```suggestion
        # odeprecated "`MacOS::CLT.version` on Linux"
```

File: Library/Homebrew/os/linux.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:111]:
```suggestion
        # odeprecated "`MacOS::CLT.installed?` on Linux"
```

**Review by MikeMcQuaid — COMMENTED** [review]
Should be good to merge after these.

**Comment by Rylan12** [comment]
I've added a rubocop to ensure that `MacOS` references in formulae only occur inside `on_macos` blocks, `on_{macos_version}` blocks, `def install`, `def post_install`, `test` blocks, and `service` blocks. This, in conjunction with the deprecations (now commented out) should help guide people away from `MacOS`.

The only tricky thing is that `brew readall` does not reliably catch issues here, since the deprecation warning will only appear when loading the formula on a Linux machine. To still show the deprecations when simulating Linux on macOS, I added deprecations to the macOS implementations as well that only trigger when simulating. I've added these as comments in b7532a86e14c7879f2815ae4c0a85e3daf7aa2ac and can remove if we don't want it.

**Comment by Bo98** [comment]
> The only tricky thing is that brew readall does not reliably catch issues here, since the deprecation warning will only appear when loading the formula on a Linux machine.

We run that usually on Ubuntu 22.04 anyway.

**Review by MikeMcQuaid — APPROVED** [review]
Great work here, thanks again @Rylan12!

**Pull request merged**
